backend/src/main/java/com/bakdata/conquery/models/index/MapInternToExternMapper.java

```diff
@@ -0,0 +1,69 @@
+package com.bakdata.conquery.models.index;
+
+
+import java.net.URL;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import com.bakdata.conquery.io.cps.CPSType;
+import com.fasterxml.jackson.annotation.JacksonInject;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.OptBoolean;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+@CPSType(id = "CSV_MAP", base = InternToExternMapper.class)
+@RequiredArgsConstructor
+@ToString(onlyExplicitlyIncluded = true)
+public class MapInternToExternMapper implements InternToExternMapper {
+
+
+	// We inject the service as a non-final property so, jackson will never try to create a serializer for it (in contrast to constructor injection)
+	@JsonIgnore
+	@JacksonInject(useInput = OptBoolean.FALSE)
+	private MapIndexService mapIndex;
+
+	@Getter
+	@ToString.Include
+	private final URL csv;
+	@Getter
+	@ToString.Include
+	private final String internalColumn;
+	@Getter
+	@ToString.Include
+	private final String externalTemplate;
+
+
+	//Manager only
+	@JsonIgnore
+	private CompletableFuture<Map<String, String>> int2ext = null;
+
+
+	@Override
+	public void init() {
+		if (mapIndex == null) {
+			// Todo ensure we are on a Worker
+			return;
+		}
+		// Manager
+		int2ext = mapIndex.getMapping(csv, internalColumn, externalTemplate);
+	}
+
+
+	@Override
+	public String external(String internalValue) {
+		try {
+			return int2ext.get(1, TimeUnit.MINUTES).getOrDefault(internalValue, "");
```

Comment:
Heißt das, dass wenn eine CSV gerendert werden soll erst die CSV geparsed/rendert wird? Das wird nicht in unter einer Minute passieren bei LANR/BSNR/PZN, das müsste schon davor passiert sein

Reply:
Das geht so:
Du lädst ein Concept mit einem mapped Select hoch. Dann wird das Concept dem Store hinzugefügt und dabei wird vom select die Init Methode getriggert. Das stößt asynchron das Parsen der CSV an, sodass die nächsten Concepts zum Store hinzugefügt werden können (und potentiell andere Parsings angeworfen werden).

Das passiert wenn importiert wird bzw. ein Storage geladen wird.

Sollte jetzt eine Anfrage eingehen und ein Ergebnis angefordert werden, dann würde sich das verzögern, weil das Parsen vielleicht noch im Gange ist, aber sobald das durch wird nicht mehr gewartet.

Das parsen wird also nicht durch ein Rendern sondern durch das laden in den Store getriggert.

Reply:
Ok, das macht Sinn.

Was passiert aktuell im Fehlerfall? Also wenn zB die Datei nicht existiert?

Reply:
Dann sollte im CacheLoader eine IOException fliegen. die wird gelogged und weiter geworfen. Die exeception sollte wiederum auftauchen, wenn ein Mapping zu bei der result Generierung angefordert wird. Dann wird im Ergebnis kein Mapping geprintet sondern "". 

Um das Problem zu lösen, sollte die Referenz datei bereit gestellt werden und das Konzept erneut eingeladen werden. Ist das okay so oder hast du an eine bessere Recovery gedacht?

Reply:
> Ist das okay so oder hast du an eine bessere Recovery gedacht?

Denke es reicht, wobei es mir besser gefallen würde den Originalwert dann mindestens zu printen. So kann man den Select zumindest noch nutzen

Reply:
Hatte ich auch überlegt, aber ich dachte dass das zu Verwirrungen führen kann, denn der Originalwert sollte über ein anderes Select (das nicht mapped) verfügbar sein.

Angenommen das Tabellen Ergebnis soll weiter verarbeitet werden und es wird von diesem Ausgabewert eine Adresse erwartet würde ich es komisch finden, wenn hier auf einmal etwas steht das keine Adresse ist. Da finde ich `""` besser um aufzuzeigen, dass es kein gerade möglich ist

Reply:
Alternativ, kann man evtl schon beim resolve schritt testen ob die mappings bereit sind und die Anfrage failen lassen, mit entsprechender Message? Stelle es mir etwas umständlich vor, aber wäre zumindest die beste Maßnahme gegen Inkonsistenz